src/core/desktopentrymonitor.cpp

```diff
@@ -0,0 +1,188 @@
+	QDir dir(path);
+
+	// Check if directory still exists - handle removal/unmounting
+	if (!dir.exists()) {
+		qCDebug(logDesktopMonitor) << "Directory no longer exists, cleaning up:" << path;
+		this->watcher->removePath(path);
+
+		// Remove all watched files from this directory and its subdirectories
+		for (auto it = this->watchedFiles.begin(); it != this->watchedFiles.end();) {
+			const QString& watchedFile = *it;
+			if (watchedFile.startsWith(path + "/") || watchedFile == path) {
+				this->watcher->removePath(watchedFile);
+				this->fileTimestamps.remove(watchedFile);
+				this->queueChange(ChangeEvent::Removed, watchedFile);
+				qCDebug(logDesktopMonitor) << "Removed file due to directory deletion:" << watchedFile;
+				it = this->watchedFiles.erase(it);
+			} else {
+				++it;
+			}
+		}
+		return;
+	}
+
+	QList<QString> currentFiles = dir.entryList({"*.desktop"}, QDir::Files);
+
+	// Check for new files
+	for (const QString& file: currentFiles) {
+		QString fullPath = dir.absoluteFilePath(file);
+		if (!this->watchedFiles.contains(fullPath)) {
+			if (this->watcher->addPath(fullPath)) {
+				this->fileTimestamps[fullPath] = QFileInfo(fullPath).lastModified();
+				this->watchedFiles.insert(fullPath);
+				this->queueChange(ChangeEvent::Added, fullPath);
+				qCDebug(logDesktopMonitor) << "New desktop file detected:" << fullPath;
+			}
+		}
+	}
+
+	// Check for deleted files
+	for (auto it = this->watchedFiles.begin(); it != this->watchedFiles.end();) {
+		const QString& watchedFile = *it;
+		if (QFileInfo(watchedFile).dir().absolutePath() == path && !QFile::exists(watchedFile)) {
+			this->watcher->removePath(watchedFile);
+			this->fileTimestamps.remove(watchedFile);
+			this->queueChange(ChangeEvent::Removed, watchedFile);
+			qCDebug(logDesktopMonitor) << "Desktop file removed:" << watchedFile;
+			it = this->watchedFiles.erase(it);
+		} else {
+			++it;
+		}
+	}
+
+	// Check for new subdirectories
+	QList<QString> subdirs = dir.entryList(QDir::Dirs | QDir::NoDotAndDotDot);
+	for (const QString& subdir: subdirs) {
+		QString subdirPath = dir.absoluteFilePath(subdir);
+		if (!this->watcher->directories().contains(subdirPath)) {
+			this->scanAndWatch(subdirPath);
+		}
+	}
+}
+
+void DesktopEntryMonitor::onFileChanged(const QString& path) {
+	if (QFileInfo(path).suffix().toLower() != "desktop") return;
+
+	QFileInfo info(path);
+	if (info.exists()) {
+		QDateTime currentModified = info.lastModified();
+		if (this->fileTimestamps.value(path) != currentModified) {
+			this->fileTimestamps[path] = currentModified;
+			this->queueChange(ChangeEvent::Modified, path);
+			qCDebug(logDesktopMonitor) << "Desktop file modified:" << path;
+		}
+	} else {
+		// File was deleted
+		this->watcher->removePath(path);
+		this->fileTimestamps.remove(path);
+		this->watchedFiles.remove(path);
+		this->queueChange(ChangeEvent::Removed, path);
+		qCDebug(logDesktopMonitor) << "Desktop file removed:" << path;
+	}
+}
```

Comment:
Timestamps cannot be relied on. We shouldn't track or use them.

Reply:
Stale leftover, I was initially watching .desktop files but decided its better to just watch the directories and re-populate on events.